text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+[`_BitInt`]: https://en.cppreference.com/w/c/language/arithmetic_types
+[`__int128`]: https://gitlab.com/x86-psABIs/x86-64-ABI/-/issues/11
+
+## Limits on `N`
+
+There are two primary limits that restrict how large `N` can be:
+
+1. All allocations in Rust are limited to `isize::MAX` bytes.
+2. Most integer methods and constants use `u32` when counting bits.
+
+The first restriction doesn't matter since `isize::MAX` bytes is `isize::MAX * 8` bits, which is larger than `usize::MAX = isize::MAX * 2 + 1` bits.
+
+However, the second restriction is somewhat significant: for systems where `usize::MAX > u32::MAX`, we are still effectively restricted to `N <= u32::MAX` unless we wish to change these APIs. We can treat this as effectively a post-monomorphisation error similar to the error you might get when adding very large arrays inside your type; it's unlikely that someone might encounter them, but they do exist and have to be accounted for.
+
+It's worth noting that `u32::MAX` bits is equivalent to 0.5 GiB, and thus no integer in Rust will be able to be larger than this amount. This is seen as acceptable because at that size, people can just use their own big-integer types. For now, adding a dedicated big-integer type to the standard library is left as a potential future change.
+
+The compiler should be allowed to restrict `N` even further, maybe even as low as `u16::MAX`, due to other restrictions that may apply. For example, the LLVM backend currently only allows integers with widths up to `u<23>::MAX` (not a typo; 23, not 32). On 16-bit targets, using `usize` further restricts these integers to `u16::MAX` bits.
+
+While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.
+
+[`generic_const_exprs`]: https://github.com/rust-lang/rust/issues/76560
+
+## Standard library
+
+The existing macro-based implementation for `uN` and `iN` should be changed to implement for only `u<N>`, `i<N>`, `usize`, and `isize` instead; this has already been implemented in a mostly-generic way and should work as expected.
+
+Unfortunately, there are a couple things that will have to remain implemented only for the existing powers of two due to the lack of constant bounds and complex const generics, namely:
+
+* `From` and `TryFrom` implementations (requires const-generic bounds)
+* `from_*e_bytes` and `to_*e_bytes` methods (requires [`generic_const_exprs`])
+
+Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.
```

Comment:
> One thing to emphasize here: getting `u128` to work was a huge endeavour, and bigger ones will be even harder for things like division -- even for 128-bit it calls out to a specific symbol for that.

LLVM has a pass specifically for expanding large divisions into a loop that doesn't use a libcall, so that shouldn't really be an issue though libcalls can still be added if you want something faster: https://github.com/llvm/llvm-project/commit/3e39b271016837425038e55e57c39e5e9b289975

Reply:
as part of clang gaining support for `_BitInt(N)` where `N > 128`, basically all the work to make it work has already been done in LLVM. Div/Rem was the last missing piece and that was added in 2022.

Reply:
Clang still limit `_BitInt(N)` to N <= 128 on quite a few targets: https://gcc.godbolt.org/z/8P3sMjavs

Reply:
i think that's merely because they haven't got around to defining the ABI, but it all works afaict: https://llvm.godbolt.org/z/88K3ox7bh

Reply:
Also worth mentioning that having `N > 128` be a post-monomorphisation error was seen as one of the biggest downsides to the previous RFC, and that this would cause more of a headache than just trying to make it work in general.

Reply:
That's a good start. However, as long as Clang isn't shipping it, the people working on Clang aren't discovering and fixing any bugs specific to those platforms. The div/rem lowering happens in LLVM IR so it's hopefully pretty target-independent, but most other operations are still legalized later in the backends. That includes the operations the div/rem lowering relies on, but also any other LLVM intrinsics that the standard library uses or may want to use in the future.

Reply:
I'm kind of relying a lot on the fact that even though not everyone is using `_BitInt(N)` right now, by the time we actually would be stabilising this RFC, LLVM would be a lot more robust in that regard. Kind of a role reversal from what happened with 128-bit integers: back then, we were really pushing LLVM to have better support, and C benefited from that, but now, C pushing LLVM to have better support will benefit Rust instead.

Reply:
As you say, this can be revisited later, but note that there's no guarantee that Clang will ever support `_BitInt(129)` or larger on any particular target. The C standard only requires `BITINT_MAXWIDTH >= ULLONG_WIDTH`. If some target keeps it at 128 for long enough, it could become entrenched enough that nobody wants to risk increasing it (e.g., imagine people putting stuff like `char bits[BITINT_MAXWIDTH / 8];` in some headers).

Reply:
I actually had no idea that was how the standard worked, but I shouldn't really be surprised, considering how it's C. :/